packetbeat/protos/smtp/parser.go

```diff
@@ -0,0 +1,261 @@
+	body    common.NetString
+
+	// Response
+	statusCode    int
+	statusPhrases []common.NetString
+
+	raw []byte
+	// list element use by 'transactions' for correlation
+	next *message
+}
+
+// Error code if stream exceeds max allowed size on append.
+var (
+	constCRLF = []byte("\r\n")
+	constEOD  = []byte(".\r\n")
+	// Responses are at least len("XXX\r\n") in size
+	constMinRespSize  = 5
+	constPhraseOffset = 4
+	constStatusSize   = 3
+	constCRLFSize     = 2
+
+	ErrStreamTooLarge = errors.New("Stream data too large")
+)
+
+func (p *parser) init(
+	cfg *parserConfig,
+	pub *transPub,
+	onMessage func(*message) error,
+) {
+	*p = parser{
+		buf:       streambuf.Buffer{},
+		config:    cfg,
+		pub:       pub,
+		onMessage: onMessage,
+	}
+}
+
+func (p *parser) append(data []byte) error {
+	_, err := p.buf.Write(data)
+	if err != nil {
+		return err
+	}
+
+	if p.config.maxBytes > 0 && p.buf.Total() > p.config.maxBytes {
+		return ErrStreamTooLarge
+	}
+	return nil
+}
+
+func (p *parser) feed(ts time.Time, data []byte) error {
+	if err := p.append(data); err != nil {
+		return err
+	}
+
+	for p.buf.Total() > 0 {
+		if p.message == nil {
+			// allocate new message object to be used by parser with current timestamp
+			p.message = p.newMessage(ts)
+		}
+
+		msg, err := p.parse()
+		if err != nil {
+			return err
+		}
+		if msg == nil {
+			break // wait for more data
+		}
+
+		// reset buffer and message -> handle next message in buffer
+		p.buf.Reset()
+		p.message = nil
+
+		// call message handler callback
+		if err := p.onMessage(msg); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *parser) newMessage(ts time.Time) *message {
+	return &message{
+		Message: applayer.Message{
+			Ts: ts,
+		},
+	}
+}
+
+// SMTP server's reply message is separated from the 3-digit status
+// code with either a space or a hyphen. The latter indicates that
+// more lines are to follow.
+func (*parser) isResponseComplete(byts []byte) bool {
+	switch byts[constStatusSize] {
+	case ' ':
+		return true
+	case '-':
+		return false
+	default:
+		debugf("Failed to understand SMTP status code: %s",
+			string(byts[:len(byts)-constCRLFSize]))
+	}
+
+	return true
+}
+
+func (p *parser) parsePayload() error {
+	if !p.pub.sendDataHeaders && !p.pub.sendDataBody {
+		return nil
+	}
+
+	m := p.message
+
+	// ".\r\n" is not part of the payload
+	parseTo := len(m.raw) - len(constEOD)
+
+	payload, err := mail.ReadMessage(bytes.NewReader(m.raw[:parseTo]))
+	if err != nil {
+		return err
+	}
+
+	if m.headers == nil {
+		m.headers = make(map[string]common.NetString)
+	}
+
+	for k := range payload.Header {
+		m.headers[k] = common.NetString(payload.Header.Get(k))
+	}
+
+	if body, err := ioutil.ReadAll(payload.Body); err != nil {
+		return err
+	} else {
+		m.body = common.NetString(body)
+	}
+
+	return nil
+}
+
+func (p *parser) parse() (*message, error) {
+	m := p.message
+
+	for {
+		var err error
+		var code int
+
+		byts, err := p.buf.CollectUntil(constCRLF)
```

Comment:
minor one, but `byts` is somewhat weird. How about one of `buf`, `raw`, `data`, `content`?